### DISCUSSION TITLE
Parameter substitution fails when statement has a postgres style cast

### DISCUSSION THREAD

**Discussion by amir-bio** [sqlalchemy/sqlalchemy#11310] [Usage Questions]
### Describe the bug

I've come across an issue where the parameters within my query are not getting substituted correctly.

Small example showing the issue:
```
import sqlalchemy
sql_str = "SELECT CAST(:param_1::VARCHAR[])"
query = sqlalchemy.text(sql_str)
query = query.bindparams(param_1=["A"])
compiled_query = query.compile(dialect=sqlalchemy.dialects.postgresql.dialect())
```
This raises an exception
```
ArgumentError: This text() construct doesn't define a bound parameter named 'param_1'
```

So it looks like there's some issue with detecting parameters when `::` casting is used

Using pyform paramstyle also doesn't work, i.e. the following also doesn't work either`"SELECT CAST(%(param_1)s::VARCHAR[])"`

However doing the cast differently does work, `SELECT CAST(:param_1 ASVARCHAR[])`

In my scenario, I'm caching the queries and reusing them, so need to be able to substitute these parameters later. I also can not control the cast style.

I've been unable to find any workarounds, so would appreciate any help if there are workarounds or knowing if this is a bug which needs to be fixed in an upcoming release

### SQLAlchemy Version in Use

1.4.52

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

postgres:14.4

### Python Version

3.12

### Operating system

Linux and OSX

### To Reproduce

```python
import sqlalchemy
sql_str = "SELECT CAST(:param_1::VARCHAR[])"
query = sqlalchemy.text(sql_str)
query = query.bindparams(param_1=["A"])
compiled_query = query.compile(dialect=sqlalchemy.dialects.postgresql.dialect())
```

### Error

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /usr/local/lib/python3.12/site-packages/sqlalchemy/sql/elements.py:1683, in TextClause.bindparams(self, *binds, **names_to_values)
   1682 try:
-> 1683     existing = new_params[key]
   1684 except KeyError as err:

KeyError: 'param_1'

The above exception was the direct cause of the following exception:

ArgumentError                             Traceback (most recent call last)
Cell In [1], line 4
      2 sql_str = "SELECT CAST(:param_1::VARCHAR[])"
      3 query = sqlalchemy.text(sql_str)
----> 4 query = query.bindparams(param_1=["A"])
      5 compiled_query = query.compile(dialect=sqlalchemy.dialects.postgresql.dialect())

File <string>:2, in bindparams(self, *binds, **names_to_values)

File /usr/local/lib/python3.12/site-packages/sqlalchemy/sql/base.py:47, in _generative(fn, *args, **kw)
     44 """Mark a method as generative."""
     46 self = args[0]._generate()
---> 47 fn(self, *args[1:], **kw)
     48 return self

File /usr/local/lib/python3.12/site-packages/sqlalchemy/sql/elements.py:1685, in TextClause.bindparams(self, *binds, **names_to_values)
   1683     existing = new_params[key]
   1684 except KeyError as err:
-> 1685     util.raise_(
   1686         exc.ArgumentError(
   1687             "This text() construct doesn't define a "
   1688             "bound parameter named %r" % key
   1689         ),
   1690         replace_context=err,
   1691     )
   1692 else:
   1693     new_params[key] = existing._with_value(value)

File /usr/local/lib/python3.12/site-packages/sqlalchemy/util/compat.py:182, in raise_(***failed resolving arguments***)
    179     exception.__cause__ = replace_context
    181 try:
--> 182     raise exception
    183 finally:
    184     # credit to
    185     # https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/
    186     # as the __traceback__ object creates a cycle
    187     del exception, replace_context, from_, with_traceback

ArgumentError: This text() construct doesn't define a bound parameter named 'param_1'

```

**Comment by CaselIT** [comment]
Hi,

That's expected. You need to escape the colons. You find that in the documentation https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.text

**Comment by amir-bio** [comment]
Hello @CaselIT , thank you for looking into this issue!

In my case, I'm generating the query in this way
```
compiled_without_binds = query.compile(
   conn.engine, dialect=sqlalchemy.dialects.postgresql.dialect(paramstyle="named")
)
```
 using sqlalchmey for this generation as well.
 
 Later in my code, `str(compiled_without_binds)` along side `compiled_without_binds.params` is cached and reused
 
 So if the colon is meant to be escaped, why is it not being escaped when compiling the query? Is there some other flag which needs to be passed in to escape it?

**Comment by CaselIT** [comment]
The escaping is only for `text()`, not for the compilation.
Also note that since 1.4 sqlalchemy is already caching the queries.

**Comment by amir-bio** [comment]
@CaselIT I see. I think I've not explained my setup properly so there's a bit of misunderstanding.
 
In my case, I'm caching the query at the application level because the query is dynamically generated and it's a very expensive operation to generate them. I then reuse these generated queries.

I'm producing the SQL string and params to cache using the snippet I shared.
I then use `sqlalchemy.text` to convert the query string into a sqlalchemy object which is then executed.
So I have something like
```
raw_query, param = get_query_from_cache(key)
if not raw_query:
    # generate query
    query = ....
    compiled_without_binds = query.compile(
        conn.engine, dialect=sqlalchemy.dialects.postgresql.dialect(paramstyle="named")
    )
    add_query_to_cache(key, query=str(compiled_without_binds), params=compiled_without_binds.params)
else:
    query = sqlalchemy.text(raw_query)

conn.execute(query, **params)
```
    
It sounds like maybe the way I'm going about caching the queries is not correct? (I've not been able to pickle the query objects as it can contain session info and it's not pickleable by default)

I've also noticed the double cast I'm seeing `CAST(:param_1::VARCHAR[] AS VARCHAR[])` is generated by sqlalchemy when I compile a query with a single cast but set the dialect to postgres

Thank you again for your help with this issue 🙏

**Comment by CaselIT** [comment]
Regarding caching, you can probably cache the generated select using your key instead of the compiled string (I'm assuming query is a select here).
Performance should be similar (caching the string will be faster, but it's probably not that much faster compared to caching the select).

In your use case the issue your are seeing is that you are trying to re-use the statement generated by the compiler in `text`. To do that with postgresql you will need to escape the cast colons as mentioned.
Solutions off the top of my head:
- use named paramstyle like you are doing but do `query_string.replace('::', '\:\:')`
- use the default naming style of the driver and use `exec_driver_sql(query_string, ...)` instead of `execute(text(query_string),...)`. There is no need to add `text()` here since `exec_driver_sql` takes a string directly. It may require some handling of the params though, but for psycopg2 it may not be necessary (and in any case it should be possible to do using the result of `compile`)
